### DISCUSSION TITLE
RabbitMQ AMQP port is random

### DISCUSSION THREAD

**Discussion by cobar79** [testcontainers/testcontainers-java#7721] [Q&A] [answered]
I am having issues in setting the spring.rabbitmq.port setting based on my test container. My development environment works find starting it ```docker run --name RogerRabbit --rm -p 15672:15672 -p 5672:5672 rabbitmq:3-management``` I see 2 sets of ports in Docker and spring.rabbitmq.port=5672 works just fine meaning the spring config connects and listens for actuator refresh events.

When I convert it to a TestContainer
```
    @Container
    public static final RabbitMQContainer RABBIT_MQ_CONTAINER =
        new RabbitMQContainer("rabbitmq:3-management")
            .withCreateContainerCmdModifier(cmd -> cmd.withName("RabbitMQ-Tester"))
            .withExposedPorts(5672, 15672)
            .withEnv("RABBITMQ_NODE_PORT", "5672")
            .withLogConsumer(new Slf4jLogConsumer(log));
```
I get a random port: RABBIT CONTAINER HOST localhost, AMQP 62883 and the spring context fails to connect using 5672

**Comment by eddumelendez** [comment]
I highly recommend reading the [Getting started with Testcontainers](https://testcontainers.com/getting-started/). If you are using spring boot then before 3.1 you can use `@DynamicPropertySource` to set the addresses or host and port with the container values. Or, if using spring boot 3.1 then just annotate the RabbitMQContainer with `@ServiceConnection` and spring boot will do it for you.

**Reply by cobar79** [reply]
Apologies for not adding the path I went down. I am starting 3 Containers (localstack, keycloak, rabbitMQ), two of which generated random ports (localstack, keycloak). I did hike down the @DynamicPropertySource path. However, the annotation is processed before the container is started and therefore you receive a ```IllegalStateException: Mapped port can only be obtained after the container is started```. If you stop the debugger at this point, you can see the only container that has started is the Lifecycle container. So I went back to defining the ports,   ``` KeycloakContainer.withExposedPorts(9181, 8080)```
 and left the properties static. ```spring.security.oauth2.resourceserver.jwt.issuer-uri=http://localhost:9181/realms/test-realm```

For the LocalStack container, I grabbed the dynamic port in @BeforeAll and put it in a Property class. Then moved the AWS Client connection logic from a static build in a @Configuration class to the service and added the loclstack url to the normal AWS connection logic.

I was also getting annoying connection exceptions from spring cloud bus implemented with RabbitMQ, even though I have all that turned off for tests.
```
spring.cloud.config.enabled=false
spring.cloud.discovery.enabled=false
spring.cloud.bus.enabled=false
rabbitmq.autoStarting=false
```
The rabbitMQ exceptions ```org.springframework.amqp.AmqpConnectException: java.net.ConnectException: Connection refused: no further information``` are just an annoyance since the test still function.
So I decided to start up a container and mimic the ```docker run --name RogerRabbit --rm -p 15672:15672 -p 5672:5672 rabbitmq:3-management``` we use in development environment. Hence this dialogue. Not sure how to mimic what is done in development with TestContainers.  

So I am just looking how to set the amqp port on the container so it is not random ``` RABBIT CONTAINER AMQP PORT 63239```
I did fine some RabbitMQ documentation and attempted ```RabbitMQContainer.withEnv("RABBITMQ_NODE_PORT", "5672")``` but I still get the connection exception.

**Reply by eddumelendez** [reply]
It's not clear for me how Testcontainers is used and how it Spring Boot/Cloud it is setup to work with Testcontainers. Are you starting manually the containers? using JUnit 5 integration? What do you mean by lifecycle container? Take into account that the word container can mean different things in spring and testcontainers contexts.

**Reply by cobar79** [reply, marked as answer]
JUnit 5 5.9.3
Testcontainers 1.19.1
Manually starting? yes
Lifecycle container? @Testcontainers()
![image](https://github.com/testcontainers/testcontainers-java/assets/42838373/a8babde7-4787-44ff-b0f5-fe135fd2ca46)

The localstack container defined in an individual test class in another microservice was working with ```@Containter and @DynamicPropertySource.``` 

Well after hours of debugging I noticed when I changed from @Container to static blocks because RabbitMQ has static ports in their class so I was exposing ports and doing port bindings, the ```@DynamicPropertySource``` ended up between multiple static blocks. When I moved the ```@DynamicPropertySrouce`` after the static blocks it worked. Rookie mistake.

I since have gone back to ```@Container``` for all the containers except the RabbitMQ container.
